radix-engine/src/system/system_db_reader.rs

```diff
@@ -918,21 +924,41 @@ impl<'a, S: SubstateDatabase> SystemDatabaseReader<'a, S> {
         node_id: &NodeId,
         partition_num: PartitionNumber,
         substate_key: &SubstateKey,
-    ) -> Option<D> {
-        if let Some(tracked) = self.tracked {
-            tracked
+    ) -> Option<Option<D>> {
+        if let Some(updates) = self.state_updates {
+            updates
+                .by_node
                 .get(node_id)
-                .and_then(|tracked_node| tracked_node.tracked_partitions.get(&partition_num))
-                .and_then(|tracked_module| {
-                    tracked_module
-                        .substates
-                        .get(&M::to_db_sort_key(&substate_key))
+                .and_then(|node_updates| match node_updates {
+                    NodeStateUpdates::Delta { by_partition } => by_partition.get(&partition_num),
                 })
-                .and_then(|tracked_key| {
-                    tracked_key
-                        .substate_value
-                        .get()
-                        .map(|e| e.as_typed().unwrap())
+                .and_then(|partition_updates| match partition_updates {
+                    PartitionStateUpdates::Delta { by_substate } => {
```

Comment:
I see this logic in loads of places :).

Jakub has built [partition_updates.get_substate_change(..)](https://github.com/radixdlt/radixdlt-scrypto/pull/1740/files#diff-5c6afaf6cf6b87254ae416cb45c7a46656b915c0e49f28c33a7ee2143d93180a) to simplify this logic - and then we can return `Option<SubstateChange>` from this method, which I think is clearer.

I mean we could even add a `get_substate_change(partition_num, substate_key)` method to `NodeStateUpdates` and a `get_substate_change(node_id, partition_num, substate_key)` method to `StateUpdates` too which would simplify this method a lot.